### PR TITLE
Make secret volume plugin use secret manager

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #19188

@gmarek

### CODE REVIEW

File: pkg/controller/volume/attachdetach/attach_detach_controller.go [hunk 1]
```diff
@@ -620,3 +620,7 @@ func (adc *attachDetachController) GetHostIP() (net.IP, error) {
 func (adc *attachDetachController) GetNodeAllocatable() (v1.ResourceList, error) {
 	return v1.ResourceList{}, nil
 }
+
+func (adc *attachDetachController) GetSecretFunc() func(namespace, name string) (*v1.Secret, error) {
+	return nil
```
Comment (liggitt) [inline pkg/controller/volume/attachdetach/attach_detach_controller.go]: returns don't match signature

File: pkg/controller/volume/persistentvolume/volume_host.go [hunk 2]
```diff
@@ -80,3 +80,7 @@ func (ctrl *PersistentVolumeController) GetHostIP() (net.IP, error) {
 func (ctrl *PersistentVolumeController) GetNodeAllocatable() (v1.ResourceList, error) {
 	return v1.ResourceList{}, nil
 }
+
+func (adc *PersistentVolumeController) GetSecretFunc() func(namespace, name string) (*v1.Secret, error) {
+	return nil
```
Comment (liggitt) [inline pkg/controller/volume/persistentvolume/volume_host.go]: fix returns

File: pkg/kubelet/secret/secret_manager.go [hunk 3]
```diff
@@ -226,6 +224,11 @@ func getSecretNames(pod *v1.Pod) sets.String {
 			}
 		}
```
Comment (liggitt) [inline pkg/kubelet/secret/secret_manager.go]: also need to iterate over container EnvFrom

File: pkg/controller/volume/persistentvolume/volume_host.go (hunk 2 above)
Comment (liggitt) [inline pkg/controller/volume/persistentvolume/volume_host.go]: why wouldn't this use ctrl.kubeClient.Secrets(namespace).Get(name)?

File: pkg/volume/secret/secret.go [hunk 4]
```diff
@@ -188,12 +192,16 @@ func (b *secretVolumeMounter) SetUpAt(dir string, fsGroup *int64) error {
...
 	}
 
-	kubeClient := b.plugin.host.GetKubeClient()
-	if kubeClient == nil {
-		return fmt.Errorf("Cannot setup secret volume %v because kube client is not configured", b.volName)
+	var secret *v1.Secret
+	if b.getSecret != nil {
+		secret, err = b.getSecret(b.pod.Namespace, b.source.SecretName)
+	} else {
+		kubeClient := b.plugin.host.GetKubeClient()
```
Comment (liggitt) [inline pkg/volume/secret/secret.go]: why are we keeping this path?

Comment (yujuhong) [inline pkg/volume/secret/secret.go]: Agreed. Why can't `getSecret` function replace this? I mentioned the same thing in the previous PR (#39558) too...

File: pkg/kubelet/secret/secret_manager.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/kubelet/secret/secret_manager.go]: What is using EnvFrom ? From what I found in the code, volume secret plugin is not using them...

File: pkg/volume/secret/secret.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/volume/secret/secret.go]: I looked into this PR and I didn't see comment about it. But yeah, we can remove it.

File: pkg/kubelet/secret/secret_manager.go (hunk 3 above)
Comment (liggitt) [inline pkg/kubelet/secret/secret_manager.go]: https://github.com/kubernetes/kubernetes/pull/40195 is open to use it

**Comment by wojtek-t** [comment]
@liggitt @yujuhong - comments applied. PTAL

**Comment by yujuhong** [comment]
Looks good to me. Leaving to @liggitt and @saad-ali for the final decision.

**Comment by saad-ali** [comment]
Volume host is ugly but that's not your fault. We'll go in a clean it out at some point.

**Comment by wojtek-t** [comment]
@saad-ali - thanks a lot!

**Pull request merged**